ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -63,6 +69,71 @@ import qualified Ouroboros.Network.Server.Socket as Server
 maxTransmissionUnit :: Int64
 maxTransmissionUnit = 4 * 1440
 
+type ConnectionTable = M.Map Socket.SockAddr ConnectionTableEntry
+
+data ConnectionTableEntry = ConnectionTableEntry {
+      cteRefs           :: ![TVar IO Int]
+    , cteLocalAddresses :: !(S.Set Socket.SockAddr)
+    }
+
+addConnection
+    :: TVar IO ConnectionTable
+    -> Socket.SockAddr
+    -> Socket.SockAddr
+    -> [TVar IO Int]
+    -> IO ()
+addConnection tblVar remoteAddr localAddr refs =
+    atomically $ do
+        tbl <- readTVar tblVar
+        case M.lookup remoteAddr tbl of
+             Nothing -> do
+                 subRefs refs
+                 writeTVar tblVar $ M.insert remoteAddr (ConnectionTableEntry refs
+                         (S.singleton localAddr)) tbl
+             Just cte -> do
+                 let refs' = refs ++ cteRefs cte
+                 subRefs $ refs'
+                 let cte' = cte { cteRefs = refs'
+                                , cteLocalAddresses = S.insert localAddr (cteLocalAddresses cte)
+                                }
+                 writeTVar tblVar $ M.insert remoteAddr cte' tbl
+  where
+    subRefs = mapM_ (\a -> modifyTVar' a (\r -> r - 1))
+
+removeConnection
+    :: TVar IO ConnectionTable
+    -> Socket.SockAddr
+    -> Socket.SockAddr
+    -> IO ()
+removeConnection tblVar remoteAddr localAddr =
+    atomically $ do
+        tbl <- readTVar tblVar
+        case M.lookup remoteAddr tbl of
```

Comment:
You could use `M.adjust` to handle both cases: updating the `tbl` value, and deleting it. There's also `M.alterF` which can do effects inside the update operation.  This way we can have it done in a single lookup.

Reply:
Thanks, will try to rewrite using `M.alterF`.